lib/graph_gateway.go

```diff
@@ -315,14 +318,46 @@ func (api *GraphGateway) loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx con
 			metrics.contextAlreadyCancelledMetric.Inc()
 		}
 
-		err := api.fetcher.Fetch(ctx, path, func(resource string, reader io.Reader) error {
+		cctx, cncl := context.WithCancel(ctx)
+		defer cncl()
+		err := api.fetcher.Fetch(cctx, path, func(resource string, reader io.Reader) error {
 			cr, err := car.NewCarReader(reader)
 			if err != nil {
 				return err
 			}
+
+			blkCh := make(chan blocks.Block, 1)
+			go func() {
+				defer close(blkCh)
+				for {
+					blk, rdErr := cr.Next()
+					if rdErr != nil {
+						err = rdErr
+					}
+					blkCh <- blk
```

Comment:
the callback to `fetcher.Fetch` may be called multiple times, so i don't think we want to cancel the context in the callback

Reply:
Thanks, that makes sense not to cancel in the callback then.

Do we need to add a cancellable context in the callback though along with a select statement around this line so that it won't just hang if a block shows up after `io.ErrNoProgress` is returned and we stop reading off the channel?